### PR TITLE
[10.0] report_wkhtmltopdf_param

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
Report Wkhtmltopdf Param
=========================

This module allows you to add new parameters for a paper format which are
then forwarded to wkhtmltopdf command as arguments. To display the arguments
that wkhtmltopdf accepts go to your command line and type 'wkhtmltopdf -H'.

A commonly used parameter in Odoo is *--disable-smart-shrinking*, that will
disable the automatic resizing of the PDF when converting. This is
important when you intend to have a layout that conforms to certain alignment.
It is very common whenever you need to conform the PDF to a predefined
layoyut (e.g. checks, official forms,...).

Usage
=====

* Go to *Settings* and press 'Activate the developer mode (with assets)'

* Go to *Settings - Technical - Reports - Paper Format*

* Add additional parameters indicating the command argument name (remember to
  add prefix -- or -) and value.

Replaces https://github.com/OCA/server-tools/pull/817

### CODE REVIEW

**Comment by JordiBForgeFlow** [comment]
@lreficent @pedrobaeza @mlaitinen @dreispt @elicoidal The module has now been renamed and moved to the OCA/reporting-engine repo.

**Comment by pedrobaeza** [comment]
Agh, remove the plural!!

**Comment by jbeficent** [comment]
It is gone. only in the branch name

**Comment by pedrobaeza** [comment]
Yeah, I see it later, sorry for the noise.

**Review by LoisRForgeFlow — APPROVED** [review]
I move my approval here :+1:

**Review by pedrobaeza — APPROVED** [review]

File: report_wkhtmltopdf_param/README.rst [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+disable the automatic resizing of the PDF when converting. This is
+important when you intend to have a layout that conforms to certain alignment.
+It is very common whenever you need to conform the PDF to a predefined
+layoyut (e.g. checks, official forms,...).
+
+
+Usage
+=====
+
+# Go to *Settings* and press 'Activate the developer mode (with assets)'
```
Comment (elicoidal) [inline report_wkhtmltopdf_param/README.rst]: use "#. " and remove blank lines in the list

File: report_wkhtmltopdf_param/models/report.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+from odoo import models
+
+
+class Report(models.Model):
+    _inherit = 'report'
+
+    def _build_wkhtmltopdf_args(self, paperformat,
+                                specific_paperformat_args=None):
+        # noinspection PyUnresolvedReferences,PyProtectedMember
```
Comment (elicoidal) [inline report_wkhtmltopdf_param/models/report.py:14]: please remove the comment

File: report_wkhtmltopdf_param/models/report_paperformat.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+                        </div>
+                    </body>
+                </html>
+            """
+            contenthtml = [tuple([1, sample_html])]
+            content = self.env['report']._run_wkhtmltopdf(
+                [], [], contenthtml, False, paperformat, False, False, False)
+            if not content:
+                raise ValidationError(_(
+                    "Failed to create a PDF using the parameters provided"))
```
Comment (elicoidal) [inline report_wkhtmltopdf_param/models/report_paperformat.py]: provided parameters.

**Review by elicoidal — APPROVED** [review]
some details. LGTM

File: report_wkhtmltopdf_param/models/report.py (hunk 2 above)
Comment (JordiBForgeFlow) [reply report_wkhtmltopdf_param/models/report.py:14]: This comment should not be removed. It is used by pyCharm to ignore the warning messages. IMHO is very beneficial.

**Pull request merged**